### DISCUSSION TITLE
Adding additional context for "Save & continue later" field?

### DISCUSSION THREAD

**Discussion by rob-c-baker** [solspace/craft-freeform#730] [Questions & Answers]
Adding a save and continue field predominantly works, but I am having an issue with some details. The form I'm trying to save appears in a number of different places on our site, with a number of different URLs, these different URLs can change the appearance of the form and adjust the default field values pushed in.

I have been looking at this: https://docs.solspace.com/craft/freeform/v3/overview/save-continue-later/

Specifically for the success and email templates, I can use `token` and `key` variables in the Return URL for the save button field but I cannot seem to pass additional, dynamic variables - that I can set at form render time - that would let me identify in those templates the context where the form was filled in and therefore generate the correct URL back to the form to allow the user to continue filling it in.

Any idea how I can go about this?

_I have looked at plucking the save field out from the form at render time and trying to adjust it's properties, but the `url` property is `protected` and it cannot be set from outside the field object. Not even sure if it would take effect anyway._

**Comment by kjmartens** [comment]
Hi @rob-baker-ar,

You should be able to access the form via Twig. While I can see that the `siteUrl` seems to not work, grabbing segments individually seems to work. Would something like this work for you?

``` twig
savedFormRedirect: '/' ~ craft.app.request.segment(1) ~ '/' ~ craft.app.request.segment(2) ~ '/' ~ form.handle ~ '/?key={key}&session-token={token}',
```

**Reply by rob-c-baker** [reply]
I did notice that you can access the form object by picking through the source code. It's not mentioned in the docs.

This got me part of the way there. The rest is a bit of a pain, but it's a workaround that works!...

I had to add hidden fields to the form - a `categoryId` and a `slug` for another entry, injected into the form on the template that renders the form:

```twig
{% set form = craft.freeform.form('[form-name]', {
    savedFormRedirect: '[The URL I need to re-make in the email template]',
    savedSession: {
        key: submissionKey,
        token: submissionToken,
    },
    overrideValues: {
        categoryId: category.id|hash,
        slug: entry.slug|hash
    }
}) %}
```

...and then un-hashed in the email template i.e. :

```twig
{% set hashedCatId = form.getLayout().getFieldByHandle('categoryId').getValue() %}
{% set catId= craft.app.security.validateData(hashedCatId ) %}
```

Then each variable is validated carefully and used to re-construct the URL.

As I said, a bit of a pain but it works and should be safe!

_P.S. If I am passing the correct URL into the `craft.freeform.form()` call, then perhaps a good improvement would be to make that same URL available to the email template?_
